addons/binding/org.openhab.binding.upb/src/main/java/org/openhab/binding/upb/handler/UPBBaseHandler.java

```
@@ -0,0 +1,149 @@
+        }
+
+        MessageBuilder message = MessageBuilder.create().destination(getId()).link(isLink()).command(commandBytes);
+
+        if (command == RefreshType.REFRESH) {
+            message.priority(UPBMessage.Priority.LOW);
+        }
+
+        messageSender.sendMessage(message);
+    }
+
+    @Override
+    public void messageReceived(UPBMessage message) {
+        if (message.getDestination() != getId() && message.getSource() != getId()) {
+            return;
+        }
+
+        if (message.getControlWord().isLink() != isLink()) {
+            return;
+        }
+
+        updateStatus(ThingStatus.ONLINE);
+
+        // parse status
+        for (Channel c : getThing().getChannels()) {
+            State newState = stateConverter.convert(message, c.getUID().getId());
+
+            if (newState != null) {
+                updateState(c.getUID(), newState);
+                break;
+            }
+        }
+    }
+
+    /**
+     * Converts a {@link Command} into an array of bytes to be passed to the UPB modem.
+     * 
+     * @param channelUID the channel to get the command for
+     * @param command the command to convert to byte form
+     * @return an array of bytes to send to the UPB modem, or null if the command could not be converted
+     */
+    protected byte[] getCommandBytes(ChannelUID channelUID, Command command) {
+        byte[] commandByte = null;
+
+        if (command == OnOffType.ON) {
+            commandByte = new byte[] { UPBMessage.Command.ACTIVATE.toByte() };
+        } else if (command == OnOffType.OFF) {
+            commandByte = new byte[] { UPBMessage.Command.DEACTIVATE.toByte() };
+        } else if (command instanceof PercentType) {
+            commandByte = new byte[] { UPBMessage.Command.GOTO.toByte(), ((PercentType) command).byteValue() };
+        } else if (command == RefreshType.REFRESH && !isLink()) {
+            commandByte = new byte[] { UPBMessage.Command.REPORT_STATE.toByte() };
+        }
```

What does it mean if we do not match, should we maybe throw a runtime exception?